pyleco/core/message.py

```diff
@@ -0,0 +1,242 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+from json import JSONDecodeError
+from typing import Any, List, Optional, Self, Tuple
+
+
+from . import VERSION_B
+from .serialization import (create_header_frame, serialize_data, interpret_header, split_name,
+                            deserialize_data,
+                            )
+
+
+# Control transfer protocol
+class Message:
+    """A message in LECO protocol.
+
+    It consists of the following bytes frames, for which each an attribute exists:
+        - `version`
+        - `receiver`
+        - `sender`
+        - `header`
+        - 0 or more `payload` frames
+
+    If you do not specify a sender, the sending program shall add it itself.
+    The :attr:`data` attribute is the content of the first :attr:`payload` frame.
+    All attributes, except the official frames, are for convenience.
+    """
+
+    version: bytes = VERSION_B
+
+    def __init__(self, receiver: bytes | str, sender: bytes | str = b"",
+                 data: Optional[bytes | str | Any] = None,
+                 header: Optional[bytes] = None,
+                 conversation_id: Optional[bytes] = None,
+                 message_id: Optional[bytes] = None,
+                 message_type: Optional[bytes] = None,
+                 **kwargs) -> None:
+        self._setup_caches()
+        self.receiver = receiver if isinstance(receiver, bytes) else receiver.encode()
+        self.sender = sender if isinstance(sender, bytes) else sender.encode()
+        self.header = (create_header_frame(conversation_id=conversation_id, message_id=message_id,
+                                           message_type=message_type)
+                       if header is None else header)
+        if isinstance(data, (bytes)):
+            self.payload = [data]
+        if isinstance(data, str):
+            self.payload = [data.encode()]
+        else:
+            self._data = data
+
+    def _setup_caches(self) -> None:
+        self._payload: List[bytes] = []
+        self._header_elements: Optional[Tuple[bytes, bytes, bytes]] = None
+        self._sender_elements: Optional[Tuple[bytes, bytes]] = None
+        self._receiver_elements: Optional[Tuple[bytes, bytes]] = None
+        self._data = None
+
+    @classmethod
+    def from_frames(cls, version: bytes, receiver: bytes, sender: bytes, header: bytes,
+                    *payload: bytes) -> Self:
+        """Create a message from a frames list, for example after reading from a socket.
+
+        .. code::
+
+            frames = socket.recv_multipart()
+            message = Message.from_frames(*frames)
+        """
+        inst = cls(receiver, sender, header=header)
```

Comment:
Because the data argument only sets the first payload frame.
From the constructor, you cannot set other payload frames.

Reply:
> Because the data argument only sets the first payload frame.
From the constructor, you cannot set other payload frames.

I see. Maybe worth pointing that distinction out in the docstring or so?

Reply:
The docstring states, that the `data` attribute is the content of the first payload frame.

Reply:
oops, I must have overlooked that, sorry.